### PR TITLE
Filterrific!

### PR DESCRIPTION
**Pull request by actioninja**
## About The Pull Request

Filter refactor + In Game Filter Editor
Accessed via VV in the dropdown of atoms. "Edit Filters.

image
https://file.house/At3g.mp4
https://file.house/6fDh.mp4 (version before I added step controls)

## Why It's Good For The Game

Makes filters actually usable without unbearable turnaround times.

## Changelog
:cl:
add: New filter editor for admins + development purposes
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
Great work!

**Comment by RaveRadbury** [comment]
I really appreciate the firm warning at the top.

File: code/modules/admin/view_variables/filterrific.dm [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+/datum/filter_editor/New(atom/target)
+	src.target = target
+
+/datum/filter_editor/ui_state(mob/user)
+	return GLOB.admin_state
+
+/datum/filter_editor/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Filteriffic", "Filteriffic")
```
Comment (stylemistake) [inline code/modules/admin/view_variables/filterrific.dm:13]: Would be better if you defined this title on the window itself, via `<Window title="">` prop.

File: tgui/packages/tgui/interfaces/Filteriffic.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        <Fragment>
+          <Button
+            icon="chevron-up"
+            onClick={() => act('increase_priority', { name: name })} />
+          <Box inline mr={1}>
+            {priority}
+          </Box>
+          <Button
+            icon="chevron-down"
+            onClick={() => act('decrease_priority', { name: name })} />
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/Filteriffic.js:205]: Would it be possible to use a NumberInput here?

**Review by stylemistake — COMMENTED** [review]
Amazing tool, I love it!

**Comment by ArcaneMusic** [comment]
Based.

File: code/__HELPERS/filters.dm [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+			"render_source" = ""
+		)
+	),
+	"drop_shadow" = list(
+		"defaults" = list(
+			"x" = 1,
+			"y" = -1,
+			"size" = 1,
+			"offset" = 0,
+			"color" = "#0000007A"
```
Comment (Jared-Fogle) [inline code/__HELPERS/filters.dm:49]: Color defines, here and everywhere else. (`code\__DEFINES\color.dm`)

File: code/__HELPERS/filters.dm [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+	for(var/i in 1 to 7)
+		//This is a very baffling and strange way of doing this but I am just preserving old functionality
+		var/X
+		var/Y
+		var/rsq
+		do
+			X = 60*rand() - 30
+			Y = 60*rand() - 30
+			rsq = X*X + Y*Y
+		while(rsq<100 || rsq>900) // Yeah let's just loop infinitely due to bad luck what's the worst that could happen?
```
Comment (Jared-Fogle) [inline code/__HELPERS/filters.dm:306]: based bro

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/__HELPERS/filters.dm [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+	if(!isnull(x))
+		.["x"] = x
+	if(!isnull(y))
+		.["y"] = y
+	if(!isnull(offset))
+		.["offset"] = offset
+	if(!isnull(flags))
+		.["flags"] = flags
+
+/proc/apply_wibbly_filters(atom/in_atom, length)
```
Comment (Qustinnus) [inline code/__HELPERS/filters.dm:296]: hahaha fuck you action :') why do u do this

File: code/__HELPERS/filters.dm (hunk 4 above)
Comment (actioninja) [reply code/__HELPERS/filters.dm:306]: This code is actually in fact from the DM reference. It was not originally added by me and I tweaked it a bit but the algorithm is pretty much a copy paste from official documentation.

File: code/__HELPERS/filters.dm (hunk 5 above)
Comment (actioninja) [reply code/__HELPERS/filters.dm:296]: It's a good name.

Comment (Qustinnus) [reply code/__HELPERS/filters.dm:296]: i meant this entire proc

**Comment by Cyberboss** [comment]
Do one for matricies now so i can finish shuttle movement 😢

File: code/__HELPERS/filters.dm (hunk 5 above)
Comment (actioninja) [reply code/__HELPERS/filters.dm:296]: This proc is _slightly_ cleaned up but I swear to god the original implementation is from the dm ref

File: tgui/packages/tgui/interfaces/Filteriffic.js (hunk 2 above)
Comment (actioninja) [reply tgui/packages/tgui/interfaces/Filteriffic.js:205]: ngl I don't know if I care enough to properly get that working now. The idea was originally that the arrows determine the direction the priority moves in but then I realized they're not actually sorted.

Comment (stylemistake) [reply tgui/packages/tgui/interfaces/Filteriffic.js:205]: Mostly for the looks.

**Review by stylemistake — DISMISSED** [review]

File: code/__HELPERS/filters.dm [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+	if(!isnull(icon))
+		.["icon"] = icon
+	if(!isnull(render_source))
+		.["render_source"] = render_source
+	if(!isnull(x))
+		.["x"] = x
+	if(!isnull(y))
+		.["y"] = y
+	if(!isnull(size))
+		.["size"] = size
```
Comment (LemonInTheDark) [inline code/__HELPERS/filters.dm:182]: Please explain the logic of this to me, it seems like an utter waste

Comment (actioninja) [reply code/__HELPERS/filters.dm:182]: This is the only way of doing this.

Comment (LemonInTheDark) [reply code/__HELPERS/filters.dm:182]: that's a real shame, but you're right.

**Pull request merged**